Support observer without any data calibration

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24924 Serialization for nn.quantized.functional modules
* **#24923 Support observer without any data calibration**
* #24804 Update mapping dictionary to support functionalmodules and pooling operations
* #24803 Handle empty qconfig for functional Modules
* #24789 Work around for bias quantization for conv and linear operators

Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

raghuramank10000 committed:
Support observer without any data calibration

Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

jerryzh168 commented:
I think you might need to rebase: https://github.com/pytorch/pytorch/commit/049284e14d150f949768e87a77d443e104452f59

jerryzh168 commented:
but why do we need this?

raghuramank100 commented:
> but why do we need this?

This is to support quantize with no eval function. In this case, the observer should still return a valid scale/zero-point in calcqparams without needing any data.

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -52,20 +53,19 @@ def calculate_qparams(self):
-        if max_val == min_val:
-            scale = 1.0
-            zero_point = 0
+            warnings.warn("must run observer before calling calculate_qparams")
```
is this scriptable?

jerryzh168 reviewed (approved):
please make sure the observer is still scriptable

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)

raghuramank10000 committed:
Update on "Support observer without any data calibration"


Replace exception with warning for un initialized min/max values to support creation of quantized models without observers.

Differential Revision: [D16923660](https://our.internmc.facebook.com/intern/diff/D16923660/)
